`std::path::absolute`

ChrisDenton opened the pull request:
Implements #59117 by adding a `std::path::absolute` function that creates an absolute path without reading the filesystem. This is intended to be a drop-in replacement for `std::fs::canonicalize` in cases where it isn't necessary to resolve symlinks. It can be used on paths that don't exist or where resolving symlinks is unwanted. It can also be used to avoid circumstances where `canonicalize` might otherwise fail.

On Windows this is a wrapper around [`GetFullPathNameW`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew). On Unix it partially implements the POSIX [pathname resolution](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13) specification, except that it stops just short of actually resolving symlinks.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

the8472 commented:
`absolute(a.join(b))` will give resolve-like behavior which some other languages have, right? If so then the comment could mention that combination as a way to resolve relative paths to something other than the current working directory.

ChrisDenton commented:
Good point! But I'm slightly wary of the Windows behaviour here. I'm thinking of legacy paths such as `D:foo` which will resolve to the `D` drive's current directory if the application is run from the command prompt (or the root of `D` otherwise).

E.g.
```rust
let path = Path::new(r"C:\path").join("D:foo");
// path is "D:foo"
let absolute = absolute(path);
// absolute is like "D:\something\foo"
// actual value depends on the `D` drive's current directory.
```

the8472 commented:
Is that an issue with `join` or with `absolute`?

ChrisDenton commented:
Hm, probably with using them to implement `resolve`. The command prompt can have a different current directory for each letter from `A` to `Z` (though it may not have them all). So to capture this you'd need to pass in mappings of (optional) current directories for each drive.

`join` could work differently. E.g. it could treat `D:foo` as either `.\D:foo` or `D:\foo`. But it doesn't and if it did it may be unexpected. I guess `absolute` could error on such paths. However, I don't think it should do any thing different from the OS as the goal here is to be as close to `canonicalize` as possible.

To be clear, I'm uncertain how much of an issue this is but it does make me wary of suggesting it.

ChrisDenton commented on library/std/src/sys/windows/path.rs:
```diff
@@ -260,3 +260,19 @@ pub(crate) fn maybe_verbatim(path: &Path) -> io::Result<Vec<u16>> {
+        // SAFETY: `fill_utf16_buf` ensures the `buffer` and `size` are valid.
+        // `lpfilename` is a pointer to a null terminated string that is not
+        // invalidated until after `GetFullPathNameW` returns successfully.
+        |buffer, size| unsafe { c::GetFullPathNameW(lpfilename, size, buffer, ptr::null_mut()) },
```
Indeed it does! The std [already does exactly that](https://github.com/rust-lang/rust/blob/3b263ceb5cb89b6d53b5a03b47ec447c3a7f7765/library/std/src/sys/windows/path.rs#L223-L229). Unfortunately the current docs for [`GetFullPathNameW`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew#parameters) suggests otherwise but this was an oversight that's going to be corrected when the online docs are next rebuilt (see the [sdk-api source](https://github.com/MicrosoftDocs/sdk-api/blob/docs/sdk-api-src/content/fileapi/nf-fileapi-getfullpathnamew.md#-param-lpfilename-in)).

Mark-Simulacrum commented:
In general this seems "fine", though I don't know if it's quite the API we want. Adding it as unstable to start though seems quite reasonable; we can discuss prior to stabilization whether this hits the right point on the API spectrum. Some initial questions I might think about:

* Should we prepend the current directory, or skip that step? The user can typically join with the current directory afterwards, right?
* Is the avoidance of stripping `..` actually the right call? It does seem consistent with the docs linked, but may not be the most intuitive behavior (particularly given that Windows doesn't do that).
* Probably some other questions based on the thread here so far, you probably have a better sense than me.

Can you file a tracking issue, and then I can r+?

ChrisDenton commented:
@Mark-Simulacrum I've filed a tracking issue. I've added your points to the unresolved questions and I'll try to give my thoughts on them soon.

Mark-Simulacrum commented:
@bors r+ rollup

Thanks! Going to go ahead and approve this PR.

bors commented:
:pushpin: Commit 0660732988ad0324ebf4437eb53f82f599c3ccfd has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 0660732988ad0324ebf4437eb53f82f599c3ccfd -->
<!-- homu: {"type":"Approved","sha":"0660732988ad0324ebf4437eb53f82f599c3ccfd","approver":"Mark-Simulacrum"} -->

ehuss commented:
@bors r-

Failed in https://github.com/rust-lang/rust/pull/92785#issuecomment-1010640697

The wasm32-unknown-unknown target reuses `unix/path.rs` but itself is not unix.

Let me know if you have any questions about that or how to test it.

ChrisDenton commented:
Thanks! I got `wasm32-unknown-unknown` to build locally. The fix was fortunately simple. I also added the tracking issue number and fixed a typo while I was there.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 232f4b47cda971522528867d09eaf9cae7dfc170 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 232f4b47cda971522528867d09eaf9cae7dfc170 -->
<!-- homu: {"type":"Approved","sha":"232f4b47cda971522528867d09eaf9cae7dfc170","approver":"Mark-Simulacrum"} -->

matthiaskrgr commented:
@bors r-  rollup=iffy
failed in a rollup:
https://github.com/rust-lang/rust/pull/92842#issuecomment-1011852470

ChrisDenton commented:
Sorry. I see why that failed but that was one of the targets I thought I was testing locally. So I'll need a minute to figure out why I didn't catch this.

ChrisDenton commented:
Ok, I've for real tested with the edge case targets: `aarch64-kmc-solid_asp3`, `x86_64-fortanix-unknown-sgx`, `wasm32-unknown-unknown`. This experience has taught me not to get "clever" with scripts and to just let Rust do its thing. Sorry again.

I squished the new changes into the last commit to avoid spamming Rust's commit history.

JohnCSimon commented:
ping from triage:
@ChrisDenton What is the status of this PR?

When it's ready for review send a message containing 
`@rustbot ready` to switch back to `S-waiting-on-review`

jyn514 commented:
This would have been very helpful for https://github.com/rust-lang/rust/pull/92260.

ChrisDenton commented:
Sorry! I've been a bit distracted for the past couple of weeks so my PRs have been languishing a bit. This one should indeed be ready.

@rustbot ready

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2 -->
<!-- homu: {"type":"Approved","sha":"fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2","approver":"Mark-Simulacrum"} -->

Mark-Simulacrum commented:
@bors rollup=iffy in case there's any trouble with platforms still

bors commented:
:hourglass: Testing commit fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2 with merge f92475c05c5eaf6db9df78c9eff1d03345a960b3...
<!-- homu: {"type":"BuildStarted","head_sha":"fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2","merge_sha":"f92475c05c5eaf6db9df78c9eff1d03345a960b3"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5098658743?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/f92475c05c5eaf6db9df78c9eff1d03345a960b3/checks/5098658743/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Preparing cargo
[2022-02-07T19:54:07Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=None, patch=None
[2022-02-07T19:54:07Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=None, patch=None
[2022-02-07T19:54:07Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=None, patch=None
[2022-02-07T19:54:07Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpIfEMzV#cargo:0.29.0" "--profile" "check" "--lib" "--" "--skip-this-rustc"
[2022-02-07T19:54:07Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpC1P4QJ#cargo:0.29.0" "--release" "--lib" "--" "--skip-this-rustc"
[2022-02-07T19:54:07Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpgH2m8x#cargo:0.29.0" "--lib" "--" "--skip-this-rustc"
[2022-02-07T19:55:43Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T19:55:43Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=Some(Full), patch=None
[2022-02-07T19:55:43Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp971H3L#cargo:0.29.0" "--profile" "check" "--lib" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T19:55:56Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrFull), patch=None
---
[2022-02-07T19:56:15Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp971H3L#cargo:0.29.0" "--profile" "check" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmp971H3L/incremental-state"
Running cargo: Debug + [Full, IncrFull, IncrUnchanged, IncrPatched]
[2022-02-07T19:56:23Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T19:56:24Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=Some(Full), patch=None
[2022-02-07T19:56:24Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpKfLeDQ#cargo:0.29.0" "--lib" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T19:56:53Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T19:56:53Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpKfLeDQ#cargo:0.29.0" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpKfLeDQ/incremental-state"
[2022-02-07T19:57:29Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T19:57:29Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpKfLeDQ#cargo:0.29.0" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpKfLeDQ/incremental-state"
[2022-02-07T19:57:36Z DEBUG collector::execute] applying println to "/tmp/.tmpKfLeDQ"
[2022-02-07T19:57:36Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T19:57:36Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T19:57:36Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpKfLeDQ#cargo:0.29.0" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpKfLeDQ/incremental-state"
[2022-02-07T19:57:51Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T19:57:51Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=Some(Full), patch=None
[2022-02-07T19:57:51Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpslZPd9#cargo:0.29.0" "--release" "--lib" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T19:58:29Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T19:58:29Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T19:58:29Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpslZPd9#cargo:0.29.0" "--release" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpslZPd9/incremental-state"
[2022-02-07T19:59:10Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T19:59:10Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpslZPd9#cargo:0.29.0" "--release" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpslZPd9/incremental-state"
[2022-02-07T19:59:17Z DEBUG collector::execute] applying println to "/tmp/.tmpslZPd9"
[2022-02-07T19:59:17Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T19:59:17Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T19:59:17Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpslZPd9#cargo:0.29.0" "--release" "--lib" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpslZPd9/incremental-state"
Preparing ctfe-stress-4
[2022-02-07T19:59:38Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=None, patch=None
[2022-02-07T19:59:38Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=None, patch=None
[2022-02-07T19:59:38Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=None, patch=None
---
[2022-02-07T20:00:33Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:00:33Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=Some(Full), patch=None
[2022-02-07T20:00:33Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpmiF2Zw#ctfe-stress-4:0.1.0" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:00:58Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:00:58Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpmiF2Zw#ctfe-stress-4:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpmiF2Zw/incremental-state"
[2022-02-07T20:01:27Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:01:27Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpmiF2Zw#ctfe-stress-4:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpmiF2Zw/incremental-state"
[2022-02-07T20:01:28Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:01:28Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=Some(Full), patch=None
[2022-02-07T20:01:28Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmppMCqud#ctfe-stress-4:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:01:53Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:01:53Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:01:53Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmppMCqud#ctfe-stress-4:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmppMCqud/incremental-state"
[2022-02-07T20:02:21Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:02:21Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmppMCqud#ctfe-stress-4:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmppMCqud/incremental-state"
Preparing externs
[2022-02-07T20:02:22Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=None, patch=None
[2022-02-07T20:02:22Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=None, patch=None
[2022-02-07T20:02:22Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=None, patch=None
[2022-02-07T20:02:22Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=None, patch=None
[2022-02-07T20:02:22Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp7FWDU7#externs:0.1.0" "--profile" "check" "--" "--skip-this-rustc"
[2022-02-07T20:02:22Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpljeezm#externs:0.1.0" "--release" "--" "--skip-this-rustc"
[2022-02-07T20:02:22Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpavtHNQ#externs:0.1.0" "--" "--skip-this-rustc"
[2022-02-07T20:02:23Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:02:23Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=Some(Full), patch=None
[2022-02-07T20:02:23Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpPt9PP3#externs:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:02:23Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrFull), patch=None
---
[2022-02-07T20:02:31Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpBWaqZd#inflate:0.1.0" "--release" "--" "--skip-this-rustc"
Running inflate: Check + [Full, IncrFull, IncrUnchanged, IncrPatched]
[2022-02-07T20:02:32Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:02:32Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=Some(Full), patch=None
[2022-02-07T20:02:32Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmprjWRRh#inflate:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:02:35Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:02:35Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmprjWRRh#inflate:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmprjWRRh/incremental-state"
[2022-02-07T20:02:37Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:02:37Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmprjWRRh#inflate:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmprjWRRh/incremental-state"
[2022-02-07T20:02:38Z DEBUG collector::execute] applying println to "/tmp/.tmprjWRRh"
[2022-02-07T20:02:38Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T20:02:38Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T20:02:38Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmprjWRRh#inflate:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmprjWRRh/incremental-state"
[2022-02-07T20:02:41Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:02:41Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=Some(Full), patch=None
[2022-02-07T20:02:41Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVvUO08#inflate:0.1.0" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:02:44Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:02:44Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:02:44Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVvUO08#inflate:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpVvUO08/incremental-state"
[2022-02-07T20:02:48Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:02:48Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVvUO08#inflate:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpVvUO08/incremental-state"
[2022-02-07T20:02:49Z DEBUG collector::execute] applying println to "/tmp/.tmpVvUO08"
[2022-02-07T20:02:49Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T20:02:49Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T20:02:49Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVvUO08#inflate:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpVvUO08/incremental-state"
[2022-02-07T20:02:52Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:02:52Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=Some(Full), patch=None
[2022-02-07T20:02:53Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpyEmTKb#inflate:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:02:58Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:02:58Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:02:58Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpyEmTKb#inflate:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpyEmTKb/incremental-state"
[2022-02-07T20:03:04Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:03:04Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpyEmTKb#inflate:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpyEmTKb/incremental-state"
[2022-02-07T20:03:05Z DEBUG collector::execute] applying println to "/tmp/.tmpyEmTKb"
[2022-02-07T20:03:05Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T20:03:05Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-02-07T20:03:05Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpyEmTKb#inflate:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpyEmTKb/incremental-state"
Preparing match-stress-enum
[2022-02-07T20:03:11Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=None, patch=None
[2022-02-07T20:03:11Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=None, patch=None
[2022-02-07T20:03:11Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=None, patch=None
---
[2022-02-07T20:03:11Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:03:11Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=Some(Full), patch=None
[2022-02-07T20:03:11Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpS8C1cF#match-stress-enum:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:03:14Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:14Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpS8C1cF#match-stress-enum:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpS8C1cF/incremental-state"
[2022-02-07T20:03:16Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:03:16Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpS8C1cF#match-stress-enum:0.1.0" "--profile" "check" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpS8C1cF/incremental-state"
[2022-02-07T20:03:16Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:03:16Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=Some(Full), patch=None
[2022-02-07T20:03:16Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpRDCI6u#match-stress-enum:0.1.0" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:03:18Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:18Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:18Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpRDCI6u#match-stress-enum:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpRDCI6u/incremental-state"
[2022-02-07T20:03:21Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:03:21Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpRDCI6u#match-stress-enum:0.1.0" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpRDCI6u/incremental-state"
[2022-02-07T20:03:21Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:03:21Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=Some(Full), patch=None
[2022-02-07T20:03:22Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpCh2xjK#match-stress-enum:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:03:24Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:24Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:24Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpCh2xjK#match-stress-enum:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpCh2xjK/incremental-state"
[2022-02-07T20:03:26Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:03:26Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpCh2xjK#match-stress-enum:0.1.0" "--release" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpCh2xjK/incremental-state"
Preparing token-stream-stress
[2022-02-07T20:03:27Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=None, patch=None
[2022-02-07T20:03:27Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=None, patch=None
[2022-02-07T20:03:27Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=None, patch=None
---
[2022-02-07T20:03:28Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:03:28Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Check, run_kind=Some(Full), patch=None
[2022-02-07T20:03:28Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpJsWS2G#token-stream-stress:0.0.0" "--profile" "check" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:03:28Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:28Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpJsWS2G#token-stream-stress:0.0.0" "--profile" "check" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpJsWS2G/incremental-state"
[2022-02-07T20:03:29Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Check, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:03:29Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpJsWS2G#token-stream-stress:0.0.0" "--profile" "check" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpJsWS2G/incremental-state"
[2022-02-07T20:03:29Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:03:29Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Debug, run_kind=Some(Full), patch=None
[2022-02-07T20:03:29Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpIwtuUc#token-stream-stress:0.0.0" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:03:29Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:29Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrFull), patch=None
[2022-02-07T20:03:29Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpIwtuUc#token-stream-stress:0.0.0" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpIwtuUc/incremental-state"
[2022-02-07T20:03:30Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Debug, run_kind=Some(IncrUnchanged), patch=None
[2022-02-07T20:03:30Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpIwtuUc#token-stream-stress:0.0.0" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln" "-C" "incremental=/tmp/.tmpIwtuUc/incremental-state"
[2022-02-07T20:03:30Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-02-07T20:03:30Z INFO  collector::execute] run_rustc with incremental=false, build_kind=Opt, run_kind=Some(Full), patch=None
[2022-02-07T20:03:30Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp0iD0nB#token-stream-stress:0.0.0" "--release" "--bin" "token-stream-stress" "--" "--wrap-rustc-with" "eprintln"
[2022-02-07T20:03:30Z INFO  collector::execute] run_rustc with incremental=true, build_kind=Opt, run_kind=Some(IncrFull), patch=None
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5099600137?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5099600137?check_suite_focus=true","builder_name":"checks-actions"} -->

bors commented:
:hourglass: Testing commit fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2 with merge 6c5f769f8a41f18756f1d499a0eecc6bd3200de9...
<!-- homu: {"type":"BuildStarted","head_sha":"fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2","merge_sha":"6c5f769f8a41f18756f1d499a0eecc6bd3200de9"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5103561801?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5103561801?check_suite_focus=true","builder_name":"checks-actions"} -->

ehuss commented:
@bors retry

curl: (6) Could not resolve host: ci-mirrors.rust-lang.org

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5103545192?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/6c5f769f8a41f18756f1d499a0eecc6bd3200de9/checks/5103545192/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2 with merge 6f73741edc2a0bce68d11587099a7c8fceac298c...
<!-- homu: {"type":"BuildStarted","head_sha":"fd5bb455c4e48c4ac4ac9b22ac6fdcb06de5c5f2","merge_sha":"6f73741edc2a0bce68d11587099a7c8fceac298c"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5109994426?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5109994426?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-mipsel-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5109943098?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/6f73741edc2a0bce68d11587099a7c8fceac298c/checks/5109943098/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] core test:false 13.144
[RUSTC-TIMING] addr2line test:false 0.269
[RUSTC-TIMING] object test:false 2.729
[RUSTC-TIMING] gimli test:false 2.759
error[E0599]: no function or associated item named `new_const` found for struct `io::error::Error` in the current scope
     |
     |
3238 |         return Err(io::Error::new_const(
     |                               ^^^^^^^^^ function or associated item not found in `io::error::Error`
    ::: library/std/src/io/error.rs:68:1
     |
68   | pub struct Error {
68   | pub struct Error {
     | ---------------- function or associated item `new_const` not found for this
For more information about this error, try `rustc --explain E0599`.
[RUSTC-TIMING] std test:false 1.370
error: could not compile `std` due to previous error
Build completed unsuccessfully in 0:00:16
```

</details>

ChrisDenton commented:
#87869 changed the way of creating a const `io::Error`. So I've rebased this PR and switch to using `io::const_io_error!` in library/std/src/path.rs. Nothing else has changed.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5110627400?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/c17568e29344c10f80a41ec1fef576f552abec61/checks/5110627400/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
thread '<unnamed>' panicked at 'static string', library/std/src/thread/tests.rs:189:9
......................................
failures:

---- path::tests::test_unix_absolute stdout ----
thread 'path::tests::test_unix_absolute' panicked at 'assertion failed: absolute(\"\").is_err()', library/std/src/path/tests.rs:1673:5

failures:
    path::tests::test_unix_absolute


test result: FAILED. 936 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 10.31s

error: test failed, to rerun pass '-p std --lib'
```

</details>

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 81cc3afe20d994d9c7cba4e75c3eed92bf98a402 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 81cc3afe20d994d9c7cba4e75c3eed92bf98a402 -->
<!-- homu: {"type":"Approved","sha":"81cc3afe20d994d9c7cba4e75c3eed92bf98a402","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 81cc3afe20d994d9c7cba4e75c3eed92bf98a402 with merge 1f4681ad7a132755452c32a987ad0f0d075aa6aa...
<!-- homu: {"type":"BuildStarted","head_sha":"81cc3afe20d994d9c7cba4e75c3eed92bf98a402","merge_sha":"1f4681ad7a132755452c32a987ad0f0d075aa6aa"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5174248817?check_suite_focus=true)
Approved by: Mark-Simulacrum
Pushing 1f4681ad7a132755452c32a987ad0f0d075aa6aa to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5174248817?check_suite_focus=true"},"merge_sha":"1f4681ad7a132755452c32a987ad0f0d075aa6aa"} -->

rust-timer commented:
Finished benchmarking commit (1f4681ad7a132755452c32a987ad0f0d075aa6aa): [comparison url](https://perf.rust-lang.org/compare.html?start=c26fbf857f82b2d0cb919c633ba0e3a90e5b9426&end=1f4681ad7a132755452c32a987ad0f0d075aa6aa).

**Summary**: This benchmark run did not return any relevant results.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

axetroy commented on library/std/src/path/tests.rs:
```diff
@@ -1665,6 +1665,64 @@ fn test_ord() {
+    assert_eq!(absolute("//a/b/c").unwrap(), Path::new("//a/b/c"));
+    assert_eq!(absolute("///a/b/c").unwrap(), Path::new("/a/b/c"));
+    assert_eq!(absolute("/a/b/c/").unwrap(), Path::new("/a/b/c/"));
+    assert_eq!(absolute("/a/./b/../c/.././..").unwrap(), Path::new("/a/b/../c/../.."));
```
Unix test without dot prefix.

```rs
assert_eq!(absolute("./a").unwrap(), Path::new("/pwd/a"));
assert_eq!(absolute("../a").unwrap(), Path::new("/parent_pwd/a"));
```

I have try in darwin. it is wrong

ChrisDenton replied:
Thank you for reporting this. I'll add more tests and fix the dot prefix case.
